### PR TITLE
Use `fetch_*` instead of `get_event_records`

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Avoid using get_event_records in this codepath as it is less efficient.

Horrifyingly, the `getLatestMaterializationEventByPartition` callsite actually gets every single record for a given asset, when we theoretically have much more efficient options available. However, fixing this would require adding a new instance method, and this is a much simpler fix.

Side note: I feel like there should be some built-in functionality for fetching all records rather than having to do this while loop business, but oh well.

## How I Tested These Changes

Existing tests, plus set the limit to 1 and ran tests to ensure that the cursoring behavior works as expected.

### CODE REVIEW

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:280]: is there a way to enforce that all the callers have a non-null `limit` set?

it looks like the main call site without a limit is this `resolve_latestMaterializationByPartition` field.  I feel like we could probably switch this to use `get_latest_storage_id_by_partition` to fetch the ids and then another call to `fetch_materializations` using a storage_id filter...

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:316]: same here.... we don't need to do the while loop if we can guarantee a non-None limit (which might just be a graphql change).

**Review by prha — CHANGES_REQUESTED** [review]

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (gibsondan) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:280]: @prha latestMaterializationByPartition is the specific field we're trying to optimize actually - if that callsite passes in, say, 27k storage IDs, would you expect us to be able to query them 10k a time without timing out? I'm seeing timeouts when I try but i'm not sure if that's related to my test dagit setup

Comment (prha) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:280]: yeah, i'd recommend using `get_latest_storage_id_by_partition` route there, which has an optimized implementation in cloud.

Comment (gibsondan) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:280]: right - i was referring to after get_latest_storage_id_by_partition returns the 27k storage ids and we are trying to fetch them by ID - i was finding that fetching those 10k events by storage ID was timing out

Comment (prha) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:280]: oh, I see.  yeah, maybe we need to refine the signatures to handle that level of pagination?

**Comment by OwenKephart** [comment]
@prha I agree that the ideal would be to avoid the while loop, but it does seem like the limit is optional across all call sites for these things, which looked like it would be a reasonably substantial refactor.

I also noticed that the logic I was using was wrong there (I was always fetching all events even if the limit was applied), so I updated that (with even more code :/) to only do this looping behavior if the limit is None.

I also implemented your suggestion for using `get_latest_storage_id_by_partition`, but realized that this does not accept a partitionKeys parameter, so decided to only use this in cases where there are lots of partition keys provided (otherwise if you just passed in e.g. 10 partition keys, we'd still end up doing a massive query for all storage ids).

File: python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py:1013]: will partitions ever be None on line 1013 given what happens on line 1011?

Relatedly - if partitions is None, do we still need to pass in the computed partitions on line 1011 to get_asset_materializations along with the storage IDs? Or does that make the underlying sql query more complicated?

Comment (OwenKephart) [reply python_modules/dagster-graphql/dagster_graphql/schema/asset_graph.py]: whoops partitions will never be None, you're right -- it's still nice to pass in the partitions even if the user is requesting all partitions because it's possible that there are partitions in storage which don't exist in the definition (e.g. you changed partitions defs at some point)

updated the code here

**Comment by gibsondan** [comment]
for what it's worth this does not seem to be sufficient to keep the original query that prompted this line of investigation (the one with 27248 partitions) from timing otu

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:293]: specifically fetching this many materializations with the default limit (which is 10k I think?) still seems to time out at the postgres layer

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (gibsondan) [inline python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:284]: interestingly i am finding that taking these out and *just* passing in the storage IDs when you have a list of storage IDs already makes it return faster?? Which might make some sense since the other fields are redundant once you have storage IDs and might just make the query more complicated

**Review by gibsondan — DISMISSED** [review]
to you for those thoughts (side-note - don't forget to dismiss reviews, this was stuck in my queue because it was waiting on prha's review to be dismissed for whatever reason)

File: python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py
Comment (gibsondan) [reply python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py:284]: in our test harness i am seeing that query return in ~20 seconds once we take those additional filters out

**Pull request merged**
